livingpark_utils/livingpark_utils.py

```diff
@@ -89,409 +87,146 @@ def notebook_init(self) -> HTML:
-                    dest_dir = os.path.join(
-                        "inputs",
-                        f'sub-{row["PATNO"]}',
-                        f'ses-{row["EVENT_ID"]}',
-                        "anat",
-                    )
-                    os.makedirs(dest_dir, exist_ok=True)
-                    dest_file = os.path.join(dest_dir, os.path.basename(filename))
-                    os.rename(filename, dest_file)
-                    row["File name"] = dest_file
-
-        # Update file names in cohort
-        cohort["File name"] = cohort.apply(
-            lambda x: self.find_nifti_file_in_cache(
-                x["PATNO"], x["EVENT_ID"], x["Description"]
-            ),
-            axis=1,
-        )
 
-        # Create symlinks to inputs if necessary
-        if link_in_outputs:
-            for file_name in cohort["File name"].values:
-                dest_dir = os.path.dirname(file_name).replace(
-                    os.path.join(self.data_cache_path, "inputs"),
-                    os.path.join(results_path, "pre_processing"),
+            if len(missing) > 0 and fallback:
+                _, missing = fallback.get_derivative(
+                    missing, symlink=symlink, timeout=timeout
                 )
-                dest_file = os.path.join(
-                    dest_dir,
-                    os.path.basename(file_name.replace(self.data_cache_path, "")),
-                )
-                if not os.path.exists(dest_file):
-                    os.makedirs(dest_dir, exist_ok=True)
-                    os.symlink(
-                        os.path.relpath(os.path.abspath(file_name), start=dest_file),
-                        dest_file,
-                    )
+                if len(missing) > 0:
+                    return pprint(f"Missing files: {missing}")
 
-    def cohort_id(self, cohort: pd.DataFrame) -> str:
-        """Return a unique id for the cohort.
+        print("Download completed")
 
-        The id is built as the hash of the sorted list of patient ids in the cohort.
-        Since cohort_ids may be used to create file names, negative signs ('-')
-        are replaced with underscore characters ('_') since SPM crashes on file names
-        containing negative signs. Therefore, the cohort id is a string that cannot
-        be cast to an integer.
+    def preprocess(
+        self,
+    ) -> None:
+        """Handle the preprocessing of a cohort.
 
-        Parameters
-        ----------
-        cohort: pd.DataFrame
-            A Pandas DataFrame with a column named 'PATNO'.
+        Raises
+        ------
+        NotImplementedError
 
-        Returns
-        -------
-        cohort_id: string
-            A string containing the unique id of the cohort.
         """
-        return str(hash(tuple(sorted(cohort["PATNO"])))).replace("-", "_")
+        raise NotImplementedError
 
+    # TODO Refactor SPM functions in their own module.
     def write_spm_batch_files(
```

Comment:
shouldn't all the SPM stuff go under `pipeline`, in a subdir called `spm`?

Reply:
Yes, the plan is to move them in `pipeline`, either as a subdir or a module. However, I thought that doing it in a future version of refactoring would be better. It would let us merge this one for now.
What are  your thoughts on that?